File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/viewHolders/tasks/ChecklistedViewHolder.java
package com.habitrpg.android.habitica.ui.viewHolders.tasks;

import com.github.data5tream.emojilib.EmojiTextView;
import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.events.TaskSaveEvent;
import com.habitrpg.android.habitica.events.commands.TaskCheckedCommand;
import com.magicmicky.habitrpgwrapper.lib.models.tasks.ChecklistItem;
import com.magicmicky.habitrpgwrapper.lib.models.tasks.Task;

import org.greenrobot.eventbus.EventBus;

import android.content.Context;
import android.graphics.Rect;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.TouchDelegate;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.OnClick;

public abstract class ChecklistedViewHolder extends BaseTaskViewHolder implements CompoundButton.OnCheckedChangeListener {

    @BindView(R.id.checkBoxHolder)
    ViewGroup checkboxHolder;

    @BindView(R.id.checkBox)
    CheckBox checkbox;

    @BindView(R.id.checklistView)
    LinearLayout checklistView;
    @BindView(R.id.checklistSeparator)
    View checklistSeparator;
    @BindView(R.id.checklistBottomSpace)
    View checklistBottomSpace;

    @BindView(R.id.checklistIndicatorWrapper)
    ViewGroup checklistIndicatorWrapper;

    @BindView(R.id.checkListCompletedTextView)
    TextView checklistCompletedTextView;
    @BindView(R.id.checkListAllTextView)
    TextView checklistAllTextView;

    public Boolean displayChecklist;

    public ChecklistedViewHolder(View itemView) {
        super(itemView);
        checklistIndicatorWrapper.setClickable(true);
        checkbox.setOnCheckedChangeListener(this);
        expandCheckboxTouchArea(checkboxHolder, checkbox);
        this.displayChecklist = false;
    }

    @Override
    public void bindHolder(Task newTask, int position) {
        super.bindHolder(newTask, position);
        this.checkbox.setChecked(this.task.completed);
        if (this.shouldDisplayAsActive()) {
            this.checkboxHolder.setBackgroundResource(this.task.getLightTaskColor());
        } else {
            this.checkboxHolder.setBackgroundColor(this.taskGray);
        }
        this.checklistCompletedTextView.setText(String.valueOf(task.getCompletedChecklistCount()));
        this.checklistAllTextView.setText(String.valueOf(task.getChecklist().size()));

        this.checklistView.removeAllViews();
        this.setDisplayChecklist(this.displayChecklist);

        this.checklistIndicatorWrapper.setVisibility(task.checklist.size() == 0 ? View.GONE : View.VISIBLE);
        if (this.rightBorderView != null) {
            this.rightBorderView.setVisibility(task.checklist.size() == 0 ? View.VISIBLE : View.GONE);
            if (this.task.getCompleted()) {
                this.rightBorderView.setBackgroundResource(this.task.getLightTaskColor());
            } else {
                this.rightBorderView.setBackgroundColor(this.taskGray);
            }
        }
    }

    abstract public Boolean shouldDisplayAsActive();

    public void setDisplayChecklist(Boolean displayChecklist) {
        this.displayChecklist = displayChecklist;
        //This needs to be a LinearLayout, as ListViews can not be inside other ListViews.
        if (this.checklistView != null) {
            if (this.displayChecklist && this.task.checklist != null) {
                LayoutInflater layoutInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
                for (ChecklistItem item : this.task.checklist) {
                    LinearLayout itemView = (LinearLayout) layoutInflater.inflate(R.layout.checklist_item_row, this.checklistView, false);
                    CheckBox checkbox = (CheckBox) itemView.findViewById(R.id.checkBox);
                    EmojiTextView textView = (EmojiTextView) itemView.findViewById(R.id.checkedTextView);
                    // Populate the data into the template view using the data object
                    textView.setText(item.getText());
                    checkbox.setChecked(item.getCompleted());
                    checkbox.setOnCheckedChangeListener(this);
                    RelativeLayout checkboxHolder = (RelativeLayout) itemView.findViewById(R.id.checkBoxHolder);
                    expandCheckboxTouchArea(checkboxHolder, checkbox);
                    this.checklistView.addView(itemView);
                }
                this.checklistSeparator.setVisibility(View.VISIBLE);
                this.checklistView.setVisibility(View.VISIBLE);
                this.checklistBottomSpace.setVisibility(View.VISIBLE);
            } else {
                this.checklistView.removeAllViewsInLayout();
                this.checklistSeparator.setVisibility(View.GONE);
                this.checklistView.setVisibility(View.GONE);
                this.checklistBottomSpace.setVisibility(View.GONE);
            }
        }
    }

    @OnClick(R.id.checklistIndicatorWrapper)
    public void onChecklistIndicatorClicked() {
        if (this.displayChecklist != null) {
            this.setDisplayChecklist(!this.displayChecklist);
        } else {
            this.setDisplayChecklist(true);
        }
        RecyclerView recyclerView = (RecyclerView)this.checklistView.getParent().getParent();
        LinearLayoutManager layoutManager = (LinearLayoutManager)recyclerView.getLayoutManager();
        layoutManager.scrollToPositionWithOffset(this.getAdapterPosition(), 15);
    }

    public void expandCheckboxTouchArea(final View expandedView, final View checkboxView){
        expandedView.post(() -> {
            Rect rect = new Rect();
            expandedView.getHitRect(rect);
            expandedView.setTouchDelegate(new TouchDelegate(rect, checkboxView));
        });
    }

    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        if (buttonView == checkbox) {
            if (isChecked != task.getCompleted()) {
                TaskCheckedCommand event = new TaskCheckedCommand();
                event.Task = task;
                event.completed =  !task.getCompleted();

                // it needs to be changed after the event is send -> to the server
                // maybe a refactor is needed here
                EventBus.getDefault().post(event);
                task.completed =event.completed;
                task.save();

            }
        } else {
            View v = (View) buttonView.getParent();
            while (v.getParent() != this.checklistView) {
                v = (View) v.getParent();
            }
            Integer position = ((ViewGroup) v.getParent()).indexOfChild(v);
            if (task.checklist.size() > position && isChecked != task.checklist.get(position).getCompleted()) {
                TaskSaveEvent event = new TaskSaveEvent();
                task.checklist.get(position).setCompleted(isChecked);
                event.task = task;
                EventBus.getDefault().post(event);
            }
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/viewHolders/tasks/ChecklistedViewHolder.kt
package com.habitrpg.android.habitica.ui.viewHolders.tasks

import android.content.Context
import android.graphics.PorterDuff
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.helpers.GroupPlanInfoProvider
import com.habitrpg.android.habitica.models.tasks.ChecklistItem
import com.habitrpg.android.habitica.models.tasks.Task
import com.habitrpg.common.habitica.extensionsCommon.getThemeColor
import com.habitrpg.common.habitica.extensionsCommon.isUsingNightModeResources
import com.habitrpg.common.habitica.helpers.MarkdownParser
import com.habitrpg.common.habitica.helpers.setParsedMarkdown
import com.habitrpg.shared.habitica.models.responses.TaskDirection
import com.habitrpg.shared.habitica.models.tasks.TaskType
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

abstract class ChecklistedViewHolder(
    itemView: View,
    scoreTaskFunc: ((Task, TaskDirection) -> Unit),
    var scoreChecklistItemFunc: ((Task, ChecklistItem) -> Unit),
    openTaskFunc: ((Task, View) -> Unit),
    brokenTaskFunc: ((Task) -> Unit),
    assignedTextProvider: GroupPlanInfoProvider?,
) : BaseTaskViewHolder(itemView, scoreTaskFunc, openTaskFunc, brokenTaskFunc, assignedTextProvider) {
    private val checkboxHolder: ViewGroup = itemView.findViewById(R.id.checkBoxHolder)
    private val checkmarkView: ImageView = itemView.findViewById(R.id.checkmark)
    private val lockView: ImageView = itemView.findViewById(R.id.lock_view)
    private val checkboxBackground: View = itemView.findViewById(R.id.checkBoxBackground)
    private val checklistView: LinearLayout = itemView.findViewById(R.id.checklistView)
    private val checklistIndicatorWrapper: ViewGroup = itemView.findViewById(R.id.checklistIndicatorWrapper)
    private val checklistCompletedTextView: TextView = itemView.findViewById(R.id.checkListCompletedTextView)
    private val checklistAllTextView: TextView = itemView.findViewById(R.id.checkListAllTextView)
    private val checklistDivider: View = itemView.findViewById(R.id.checklistDivider)

    init {
        checklistIndicatorWrapper.isClickable = true
        checklistIndicatorWrapper.setOnClickListener { onChecklistIndicatorClicked() }
    }

    override fun bind(
        data: Task,
        position: Int,
        displayMode: String,
        ownerID: String?,
    ) {
        var completed = data.completed(userID)
        if (data.isPendingApproval) {
            completed = false
        }
        if (isLocked) {
            this.checkmarkView.visibility = View.GONE
            this.lockView.visibility = View.VISIBLE
            val icon = AppCompatResources.getDrawable(context, R.drawable.task_lock)
            icon?.setTint(ContextCompat.getColor(context, if (data.isDue == true || data.type == TaskType.TODO) data.extraExtraDarkTaskColor else R.color.text_dimmed))
            lockView.setImageDrawable(icon)
        } else {
            this.checkmarkView.visibility = if (completed) View.VISIBLE else View.GONE
            checkmarkView.drawable.setTint(ContextCompat.getColor(context, R.color.gray_400))
            this.lockView.visibility = View.GONE
        }
        this.checklistCompletedTextView.text = data.completedChecklistCount.toString()
        this.checklistAllTextView.text = data.checklist?.size.toString()

        this.checklistView.removeAllViews()
        this.updateChecklistDisplay()

        this.checklistIndicatorWrapper.visibility = if (data.checklist?.size == 0) View.GONE else View.VISIBLE
        super.bind(data, position, displayMode, ownerID)
        val regularBoxBackground = if (task?.type == TaskType.DAILY) R.drawable.daily_unchecked else R.drawable.todo_unchecked
        val completedBoxBackground = if (task?.type == TaskType.DAILY) R.drawable.daily_checked else R.drawable.todo_checked
        val inactiveBoxBackground = R.drawable.daily_inactive
        if (this.shouldDisplayAsActive(data, userID) && !data.isPendingApproval) {
            this.checkboxHolder.setBackgroundResource(data.lightTaskColor)
            checkboxBackground.setBackgroundResource(regularBoxBackground)
        } else {
            if (completed) {
                titleTextView.setTextColor(ContextCompat.getColor(context, R.color.text_quad))
                notesTextView?.setTextColor(ContextCompat.getColor(context, R.color.text_quad))
                this.checkboxHolder.setBackgroundColor(context.getThemeColor(R.attr.colorWindowBackground))
                checkboxBackground.setBackgroundResource(completedBoxBackground)
            } else {
                this.checkboxHolder.setBackgroundColor(this.taskGray)
                notesTextView?.setTextColor(ContextCompat.getColor(context, R.color.text_ternary))
                checkboxBackground.setBackgroundResource(regularBoxBackground)
                checkboxBackground.setBackgroundResource(inactiveBoxBackground)
            }
        }
    }

    abstract fun shouldDisplayAsActive(
        task: Task?,
        userID: String?,
    ): Boolean

    private fun updateChecklistDisplay() {
        // This needs to be a LinearLayout, as ListViews can not be inside other ListViews.
        if (this.shouldDisplayExpandedChecklist()) {
            val layoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as? LayoutInflater
            if (this.task?.checklist?.isValid == true) {
                checklistView.removeAllViews()
                for (item in this.task?.checklist ?: emptyList<ChecklistItem>()) {
                    val itemView = layoutInflater?.inflate(R.layout.checklist_item_row, this.checklistView, false)
                    val checkboxBackground = itemView?.findViewById<View>(R.id.checkBoxBackground)
                    if (task?.type == TaskType.TODO) {
                        checkboxBackground?.setBackgroundResource(R.drawable.round_checklist_unchecked)
                    }
                    checkboxBackground?.backgroundTintList =
                        ContextCompat.getColorStateList(
                            context,
                            (
                                if (context.isUsingNightModeResources()) {
                                    if (task?.completed(userID) == true || (task?.type == TaskType.DAILY && task?.isDue == false)) {
                                        R.color.checkbox_fill
                                    } else {
                                        task?.lightTaskColor
                                    }
                                } else {
                                    R.color.checkbox_fill
                                }
                            ) ?: R.color.checkbox_fill,
                        )
                    val textView = itemView?.findViewById<TextView>(R.id.checkedTextView)
                    // Populate the data into the template view using the data object
                    textView?.text = item.text
                    textView?.setTextColor(ContextCompat.getColor(context, if (item.completed) R.color.text_dimmed else R.color.text_secondary))
                    if (item.text != null) {
                        MainScope().launch(Dispatchers.IO) {
                            val parsedText = MarkdownParser.parseMarkdown(item.text ?: "")
                            withContext(Dispatchers.Main) {
                                textView?.setParsedMarkdown(parsedText)
                            }
                        }
                    }
                    val checkmark = itemView?.findViewById<ImageView>(R.id.checkmark)
                    checkmark?.drawable?.setTintMode(PorterDuff.Mode.SRC_ATOP)
                    checkmark?.visibility = if (item.completed) View.VISIBLE else View.GONE
                    val checkboxHolder = itemView?.findViewById<View>(R.id.checkBoxHolder) as? ViewGroup
                    checkboxHolder?.setOnClickListener { _ ->
                        task?.let { scoreChecklistItemFunc(it, item) }
                    }
                    val color =
                        ContextCompat.getColor(
                            context,
                            if (task?.completed(userID) == true || (task?.type == TaskType.DAILY && task?.isDue == false)) {
                                checkmark?.drawable?.setTint(ContextCompat.getColor(context, R.color.text_dimmed))
                                R.color.offset_background
                            } else {
                                val color = if (context.isUsingNightModeResources()) task?.extraExtraDarkTaskColor else task?.darkTaskColor
                                checkmark?.drawable?.setTint(ContextCompat.getColor(context, color ?: R.color.text_dimmed))
                                task?.extraLightTaskColor ?: R.color.offset_background
                            },
                        )
                    color.let { checkboxHolder?.setBackgroundColor(it) }
                    this.checklistView.addView(itemView)
                }
            }
            this.checklistView.visibility = View.VISIBLE
        } else {
            this.checklistView.removeAllViewsInLayout()
            this.checklistView.visibility = View.GONE
        }
    }

    protected fun setChecklistIndicatorBackgroundActive(isActive: Boolean) {
        val drawable = ContextCompat.getDrawable(context, R.drawable.checklist_indicator_background)
        if (isActive) {
            drawable?.setTint(ContextCompat.getColor(context, R.color.gray_200))
            val textColor =
                if (context.isUsingNightModeResources()) {
                    ContextCompat.getColor(context, R.color.gray_600)
                } else {
                    ContextCompat.getColor(context, R.color.gray_500)
                }
            checklistCompletedTextView.setTextColor(textColor)
            checklistAllTextView.setTextColor(textColor)
            checklistDivider.setBackgroundColor(textColor)
        } else {
            drawable?.setTint(ContextCompat.getColor(context, R.color.offset_background))
            val textColor = ContextCompat.getColor(context, R.color.text_quad)
            checklistCompletedTextView.setTextColor(textColor)
            checklistAllTextView.setTextColor(textColor)
            checklistDivider.setBackgroundColor(textColor)
        }
        drawable?.setTintMode(PorterDuff.Mode.MULTIPLY)
        checklistIndicatorWrapper.background = drawable
    }

    private fun onChecklistIndicatorClicked() {
        expandedChecklistRow = if (this.shouldDisplayExpandedChecklist()) null else bindingAdapterPosition
        if (this.shouldDisplayExpandedChecklist()) {
            val recyclerView = this.checklistView.parent.parent as? RecyclerView
            val layoutManager = recyclerView?.layoutManager as? LinearLayoutManager
            layoutManager?.scrollToPositionWithOffset(this.bindingAdapterPosition, 15)
        }
        updateChecklistDisplay()
    }

    override fun onLeftActionTouched() {
        super.onLeftActionTouched()
        if (task?.isValid == true && !isLocked) {
            onCheckedChanged(!(task?.completed(userID) ?: false))
        }
    }

    override fun onRightActionTouched() {
        super.onRightActionTouched()
        onChecklistIndicatorClicked()
    }

    private fun shouldDisplayExpandedChecklist(): Boolean {
        return expandedChecklistRow != null && bindingAdapterPosition == expandedChecklistRow
    }

    private fun onCheckedChanged(isChecked: Boolean) {
        if (task?.isValid != true) {
            return
        }
        if (isChecked != task?.completed(userID)) {
            task?.let { scoreTaskFunc(it, if (task?.completed(userID) == false) TaskDirection.UP else TaskDirection.DOWN) }
        }
    }

    override fun setDisabled(
        openTaskDisabled: Boolean,
        taskActionsDisabled: Boolean,
    ) {
        super.setDisabled(openTaskDisabled, taskActionsDisabled)
        this.checkboxHolder.isEnabled = !taskActionsDisabled
    }

    companion object {
        private var expandedChecklistRow: Int? = null
    }
}
